net/httpbody.go

```diff
@@ -0,0 +1,277 @@
+package net
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/zalando/skipper/metrics"
+)
+
+var (
+	ErrClosed  = errors.New("reader closed")
+	ErrBlocked = errors.New("blocked string match found in body")
+)
+
+const (
+	defaultReadBufferSize uint64 = 8192
+)
+
+type MaxBufferHandling int
+
+const (
+	MaxBufferBestEffort MaxBufferHandling = iota
+	MaxBufferAbort
+)
+
+type logBody struct {
+	ctx    context.Context
+	fmtstr string
+	log    func(format string, args ...interface{})
+	input  io.ReadCloser
+}
+
+func newLogBody(ctx context.Context, fmtstr string, log func(format string, args ...interface{}), rc io.ReadCloser) io.ReadCloser {
+	return &logBody{
+		ctx:    ctx,
+		fmtstr: fmtstr,
+		input:  rc,
+		log:    log,
+	}
+}
+
+func (lb *logBody) Read(p []byte) (int, error) {
+	n, err := lb.input.Read(p)
+	if n > 0 {
+		lb.log("%s%s", lb.fmtstr, p)
+	}
+	return n, err
+}
+
+func (lb *logBody) Close() error {
+	return lb.input.Close()
+}
+
+func LogBody(ctx context.Context, fmtstr string, log func(format string, args ...interface{}), rc io.ReadCloser) io.ReadCloser {
+	return newLogBody(ctx, fmtstr, log, rc)
+}
+
+type matcher struct {
+	ctx               context.Context
+	once              sync.Once
+	input             io.ReadCloser
+	f                 func([]byte) (int, error)
+	maxBufferSize     uint64
+	maxBufferHandling MaxBufferHandling
+	readBuffer        []byte
+
+	ready   *bytes.Buffer
+	pending *bytes.Buffer
+
+	metrics metrics.Metrics
+
+	err    error
+	closed bool
+}
+
+var (
+	ErrMatcherBufferFull = errors.New("matcher buffer full")
+)
+
+func newMatcher(
+	ctx context.Context,
+	input io.ReadCloser,
+	f func([]byte) (int, error),
+	maxBufferSize uint64,
+	mbh MaxBufferHandling,
+) *matcher {
+
+	rsize := defaultReadBufferSize
+	if maxBufferSize < rsize {
+		rsize = maxBufferSize
+	}
+	println("maxBufferSize:", maxBufferSize, "rsize:", rsize)
+
+	return &matcher{
+		ctx:               ctx,
+		once:              sync.Once{},
+		input:             input,
+		f:                 f,
+		maxBufferSize:     maxBufferSize,
+		maxBufferHandling: mbh,
+		readBuffer:        make([]byte, rsize),
+		pending:           bytes.NewBuffer(nil),
+		ready:             bytes.NewBuffer(nil),
+		metrics:           metrics.Default,
+	}
+}
+
+func (m *matcher) readNTimes(times int) (bool, error) {
+	var consumedInput bool
+	for i := 0; i < times; i++ {
+		n, err := m.input.Read(m.readBuffer)
+		println("readNTimes(", times, "): read n bytes:", n, "eof:", err == io.EOF)
+		n2, err2 := m.pending.Write(m.readBuffer[:n])
+		println("readNTimes(", times, "): wrote n2 bytes:", n2, "eof:", err == io.EOF)
+
+		if n > 0 {
+			consumedInput = true
+		}
+
+		if err != nil {
+			return consumedInput, err
+		}
+		if err2 != nil {
+			return consumedInput, err
```

Comment:
What error to return here: `err`, because it's propagating state of Read or `err2` because it is what happened at the Write?